src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -473,9 +473,47 @@ private static bool CheckConstraintsSingleType(TypeSymbol type, CheckConstraints
         // C# does not let you declare a type in which it would be possible for distinct base interfaces
         // to unify under some instantiations.  But such ill-formed classes can come in through
         // metadata and be instantiated in C#.  We check to see if that's happened.
-        private static bool InterfacesAreDistinct(NamedTypeSymbol type, ConsList<Symbol> basesBeingResolved)
+        private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol> basesBeingResolved)
         {
-            return !type.InterfacesNoUseSiteDiagnostics(basesBeingResolved).HasDuplicates(TypeSymbol.EqualsIgnoringDynamicComparer);
+            // PERF: avoid instantiating all interfaces here
+            //       Ex: if class implements just IEnumerable<> and IComparable<> it cannot have conflicting implementations
+            var array = type.OriginalDefinition.InterfacesNoUseSiteDiagnostics(basesBeingResolved);
+
+            switch (array.Length)
+            {
+                case 0:
+                case 1:
+                    // less than 2 interfaces
+                    return false;
+
+                case 2:
+                    if (ReferenceEquals(array[0].OriginalDefinition, array[1].OriginalDefinition))
+                    {
+                        break;
+                    }
+
+                    // two unrelated interfaces 
+                    return false;
+
+                default:
+                    var set = new HashSet<NamedTypeSymbol>(ReferenceEqualityComparer.Instance);
+                    foreach (var i in array)
+                    {
+                        if (!set.Add(i.OriginalDefinition))
+                        {
+                            goto hasRelatedInterfaces;
```

Comment:
Duplicate code could lead to something going out of sync. For example - if conditions are inverted (like in this change), then two parts of the method must be carefully updated to match each other. - that would be easy to miss.

I think goto is used appropriately here - to exit out of two enclosed constructs.